martin-mbtiles/src/mbtiles.rs

```diff
@@ -378,32 +388,102 @@ impl Mbtiles {
+        register_md5_function(&rusqlite_conn)?;
+        rusqlite_conn.execute(
+            format!("INSERT OR REPLACE INTO metadata(name, value) {select_from}").as_str(),
+            [],
+        )?;
+        Ok(())
+    }
+
+    pub async fn validate_mbtiles<T>(
+        &self,
+        integrity_check: IntegrityCheck,
+        conn: &mut T,
+    ) -> MbtResult<()>
+    where
+        for<'e> &'e mut T: SqliteExecutor<'e>,
+    {
+        // SQLite Integrity check
+        if "ok"
+            != match integrity_check {
+                IntegrityCheck::Full => query("PRAGMA integrity_check;")
+                    .fetch_one(&mut *conn)
+                    .await?
+                    .get::<String, _>(0),
+                IntegrityCheck::Quick => query("PRAGMA integrity_check;")
+                    .fetch_one(&mut *conn)
+                    .await?
+                    .get::<String, _>(0),
+                IntegrityCheck::Off => "ok".to_string(),
             }
+        {
+            return Err(FailedIntegrityCheck(self.filepath().to_string()));
+        }
+
+        let mbttype = self.detect_type(&mut *conn).await?;
+
+        if mbttype == MbtType::Flat {
+            println!(
+                "No hash column in flat table format, skipping hash-based validation steps..."
+            );
+            return Ok(());
         }
-        .to_string();
 
         let rusqlite_conn = RusqliteConnection::open(Path::new(self.filepath()))?;
         register_md5_function(&rusqlite_conn)?;
 
+        // Global hash check
+        if self
+            .get_metadata_value(&mut *conn, "global_hash")
+            .await?
+            .is_none()
+        {
+            println!(
+                "No value for 'global_hash' key found in metadata, skipping global hash validation step..."
+            );
+        } else {
+            let sql = if mbttype == MbtType::FlatWithHash {
+                "SELECT * FROM metadata 
+                          WHERE name='global_hash' 
+                                AND value!=(SELECT hex(md5_concat(tile_hash)) FROM tiles_with_hash ORDER BY zoom_level, tile_column, tile_row);"
+            } else {
+                "SELECT * FROM metadata 
+                          WHERE name='global_hash' 
+                                AND value!=(hex(md5_concat(images.tile_id)) FROM images JOIN map ON images.tile_id=map.tile_id ORDER BY zoom_level, tile_column, tile_row);"
+            }
+                .to_string();
+
+            if rusqlite_conn.prepare(&sql)?.exists(())? {
+                return Err(InvalidTileData(self.filepath().to_string()));
+            }
+        }
+
+        // Per-tile hash check
+        let sql = if mbttype == MbtType::FlatWithHash {
+            "SELECT * FROM tiles_with_hash WHERE tile_hash!=hex(md5(tile_data)) LIMIT 1;"
+        } else {
+            "SELECT * FROM images WHERE tile_id!=hex(md5(tile_data)) LIMIT 1;"
+        }
+        .to_string();
```

Comment:
you don't need this as a string because you don't modify it. In the expression `let foo = "abc";` -- `foo` is already a `&str` that you should be able to pass to `prepare`
```suggestion
        ;
```